### ISSUE TITLE
How to compile?

### ISSUE THREAD

**Issue by BigheadSMZ** [dxx-rebirth/dxx-rebirth#770]
Is there a guide anywhere that explains the entire process and workflow required to compile this? I am far from fluent in programming, and I have no idea how to troubleshoot issues or figure out exactly what I even need. Many popular projects include a compilation guide, so I was hoping this one would too. I would like to build the latest code since the version I have is quite old now.

I tried to follow the steps here:
https://github.com/dxx-rebirth/dxx-rebirth/issues/711#issuecomment-1881095981

Unfortunately I got to the "compile" steps and got this error:
```
file F:\mysys64\dxx-rebirth\SConstruct,line 4901:
	Configure(confdir = build\.sconf_temp)
scons: dxx: version: (Git.ComputedExtraVersion(describe='0.60.0-beta2-2927-g30fdb4ae403c', status=None, diffstat_HEAD=None, revparse_HEAD='archive_30fdb4ae403ce2b0da9ac48cf67dcdf35ae612c2'),)
scons: Configure: DXX-Archive.3: checking version of C++ compiler 'g++' ... 
scons: Configure: error: The system cannot find the file specified
```
Unfortunately those steps don't say anything about "g++" and looking into it has just made me more confused. I think it would be beneficial if someone who knows how to compile this project could write a guide on how to compile since there are no automated builds available anymore.

**Comment by vLKp** [comment]
The instructions you requested are in `INSTALL.markdown`, stored at the root of the repository.  Secondarily, recent outside contributions have enabled GitHub Actions to perform an automated build for Windows, Linux, and MacOS.  The Windows automated build yaml can be used as a reference for how to build on a personal Windows system.

> since there are no automated builds available anymore.

This is backwards.  Rebirth went without automated builds for longer than many other projects, but thanks to external contributions to enable GitHub Actions, it now has automated builds.

I think the label `build-failure` is not appropriate here.  Its intent is to be used when a supported configuration fails to build, usually because of an error in the code, but sometimes because the code is no longer compatible with some external library.  In your case, your configuration is not supported, because you do not have a C++ compiler.  Rebirth is a C++ program, so a working C++ compiler is a prerequisite to building the program.  This is why the configure script starts by checking the version of your compiler, and implicitly in that check, checks that you have a working compiler.

> And the errors when trying the fourth command.

This exposes an interesting edge case.  If you have no tools to compile the program, you cannot clean the results of a prior compilation.  This is unlikely to matter in practice, but fixing it might slightly speed up the clean step.

---

If you only want to play the latest code, the easiest path would be for you to use the artifact built from the automated build.  If you still want to compile the program yourself, then please read `INSTALL.markdown` for instructions, and ask if you still need help.

**Comment by vLKp** [comment]
The most recent successful automated build can be seen at https://github.com/dxx-rebirth/dxx-rebirth/actions/runs/8493176778 .  The artifacts it produced are listed at the bottom of the page, though GitHub will only allow you to fetch them if you are logged in.

https://github.com/dxx-rebirth/dxx-rebirth/actions lists all workflows, but is somewhat cluttered by the system attempting to run workflows on pull requests.  It wants my approval before running them, but when I try to give approval, nothing happens, so those records are just stuck there.  If there is a way to delete a workflow without approving it, I have not found that way.  (The GitHub Actions documentation (https://docs.github.com/en/actions/managing-workflow-runs/deleting-a-workflow-run) suggests clicking on an empty block, but that does not do anything.)

**Comment by BigheadSMZ** [comment]
Hey sorry about posting in the category that I did. I figured since I was having trouble building it, it would fit in this kind of category so that's my mistake. 

I never did figure it out, it's a bit too complex for me and I gave up shortly after posting. I did read through the page on how to build and several other pages talking about it, but it reads like it assumes you at least have some semblance of what you are doing. Like not being able to find "g++" was a problem I had no idea of how to solve. Build instructions in `INSTALL.markdown` are somewhat vague in that it tells you everything you need but not how to actually set it all up to the point where it does what you want ; if that makes sense. An ELI5 could go a long way for dummies like me that haven't even heard of "scons" or "mingw-gcc" before let alone know how to work them. I know that writing a guide that gives step-by-step instructions takes a considerable amount of time to create, so I can understand why one doesn't exist. 

I can usually educate myself using search engines, but it's hard to find any relevant information on how to "obtain" or "link" to whatever "g++" is and get it working. In my searches I've found some random executables named "g++.exe" that were supposedly compilers but I don't think they are what I was looking for. Searching for the error and whatever other errors I was getting was not yielding the correct solutions. It doesn't help my only programming experience is with simple .NET C# code and PowerShell. When I try to build stuff that doesn't build in Visual Studio with a ".sln" file I'm a lost cause. Eventually I'll try again just to see if I can eventually figure all of this out since it would be useful knowledge to have.

Anyway, I looked for automated builds but I couldn't find the correct page so thanks for the link! I found some similar pages where all the artifacts expired, maybe I was looking at older build pages or something. The top search engine result I got when looking for "dxx-rebirth latest build" was the issue mentioning outdated builds, which has dozens of posts and links all over it leading me down an unnecessary rabbit hole. It might be a good idea to put a link to the Github Actions page in the `README.md` that says something like "automated builds here". Really the only reason I tried to build it is because I basically just wanted a way to be able to try the latest builds, since the version I've been using is from 2022. So now that I know there is working automated builds, my issue is solved.

**Comment by vLKp** [comment]
Yes, some level of proficiency is assumed.  For Linux users, this is all much easier, as Linux systems have a working package manager that can deal with installing the prerequisites for you.  I recognize that a step-by-step guide would be nice, but I have avoided it both because of the tedium involved in writing it and because I no longer maintain a Windows system, so I cannot test such a guide.  I fear that giving an apparent step-by-step that may not actually work would be worse than giving a generalized guide that requires users to do some problem solving on their own.

Visual Studio support would be a nice alternative, but when I last tried it (around the time of Visual Studio 2017, if I recall correctly), the compiler kept crashing trying to parse constructs that the GNU compiler had accepted for years.  More recently, I found that I cannot get newer versions of Visual Studio because Microsoft only offers a download of a tool that will download Visual Studio (rather than offering an actual download of the program).  Even beyond that, the last version I tried broke after a few weeks of being offline.  It was _very_ determined that I would sign in to a Microsoft Account or it would not work at all.  That behavior is not acceptable, and got it uninstalled as unusable.

Yes, I think artifacts may expire after a while.  I think I publish often enough that newer artifacts will be created before old ones expire.  The `README.md` was last updated before the automated builds existed.  It is in need of other updates, too.

**Comment by BigheadSMZ** [comment]
Yeah I can understand that completely, I know that writing a comprehensive guide that is actually helpful and informative takes a great deal of work and patience. For example, [this guide](https://forums.dolphin-emu.org/Thread-how-to-install-texture-packs-custom-textures-info) I made for custom textures in Dolphin emulator has been written and rewritten many times over the years. And over time information tends to change, and the more thorough the instructions, the more that needs to be updated later. Plus in at least your case, that time is better spent working on this great project. I've been playing Descent 1 and 2 using DXX-Rebirth for many years now on and off and it's kept these amazing games alive for me. So I can greatly appreciate and thank you for the effort you have put into this project.
